VersionControlUI/src/DiffManager.cpp

```diff
@@ -0,0 +1,288 @@
+ ** OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ **
+ **********************************************************************************************************************/
+
+#include "DiffManager.h"
+#include "VersionControlUIException.h"
+
+#include "VisualizationBase/src/VisualizationManager.h"
+#include "VisualizationBase/src/items/ViewItem.h"
+#include "VisualizationBase/src/nodes/ViewItemNode.h"
+#include "VisualizationBase/src/ViewItemManager.h"
+#include "VisualizationBase/src/declarative/GridLayouter.h"
+#include "VisualizationBase/src/overlays/HighlightOverlay.h"
+
+#include "FilePersistence/src/version_control/GitRepository.h"
+#include "FilePersistence/src/simple/SimpleTextFileStore.h"
+
+#include "OOModel/src/declarations/Project.h"
+#include "OOModel/src/statements/ExpressionStatement.h"
+#include "OOModel/src/expressions/EmptyExpression.h"
+
+#include "ModelBase/src/model/AllTreeManagers.h"
+#include "ModelBase/src/model/TreeManager.h"
+#include "ModelBase/src/nodes/Reference.h"
+
+namespace VersionControlUI
+{
+
+struct ChangeWithNodes {
+		ChangeWithNodes(Model::Node* oldNode, Model::Node* newNode,
+							 FilePersistence::ChangeType changeType)
+			: oldNode_{oldNode}, newNode_{newNode}, changeType_{changeType}
+		{}
+
+		Model::Node* oldNode_;
+		Model::Node* newNode_;
+		FilePersistence::ChangeType changeType_;
+};
+
+DiffManager::DiffManager(QString oldVersion, QString newVersion, QString project, QString symbol) :
+	oldVersion_{oldVersion}, newVersion_{newVersion}, project_{project}, symbol_{symbol}
+{}
+
+void DiffManager::visualize()
+{
+	QString projectsDir = "projects/" + project_;
+
+	// get GitRepository
+	if (!FilePersistence::GitRepository::repositoryExists(projectsDir))
+	{
+		qDebug() << "no repository found at " + projectsDir;
+		return;
+	}
+
+	FilePersistence::GitRepository repository{projectsDir};
+
+	// load newer version
+	Model::TreeManager* newVersionManager = createTreeManagerFromVersion(repository, newVersion_);
+
+	// load older version
+	Model::TreeManager* oldVersionManager = createTreeManagerFromVersion(repository, oldVersion_);
+
+	Model::Reference::resolvePending();
+
+	FilePersistence::Diff diff = repository.diff(oldVersion_, newVersion_);
+
+	auto changes = diff.changes();
+
+	// detailed changes
+	QList<ChangeWithNodes> changesWithNodes;
+
+	// contains the nodes which will be drawn
+	QSet<Model::NodeIdType> changedNodesToVisualize;
+
+	for (auto change : changes.values())
+	{
+		// TODO check flags
+		if (change->isFake() || change->onlyStructureChange()) continue;
+
+		auto id = change->nodeId();
+
+		changesWithNodes.append(ChangeWithNodes{const_cast<Model::Node*>(oldVersionManager->nodeIdMap().node(id)),
+														 const_cast<Model::Node*>(newVersionManager->nodeIdMap().node(id)),
+														 change->type()});
+
+		Model::NodeIdType nodeId;
+
+		// TODO maybe use node instead of id
```

Comment:
Yes, changedNodesToVisualize and DiffManager::findChangedNode still use NodeIdType.